{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Process metadata\n",
    "\n",
    "Not all experiment ids in the metadata file have corresponding gene expression data.  This notebook checks each experiment id and returns a clean list of experiment ids that have gene expression data.  This file will be used in ```2_generate_data.ipynb```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import os\n",
    "import sys\n",
    "import glob\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import random\n",
    "\n",
    "sys.path.append(\"../../\")\n",
    "from functions import utils\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings(action='ignore')\n",
    "\n",
    "from numpy.random import seed\n",
    "randomState = 123\n",
    "seed(randomState)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in config variables\n",
    "config_file = os.path.abspath(os.path.join(os.getcwd(),\"../../configs\", \"config_Pa_experiment.tsv\"))\n",
    "params = utils.read_config(config_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load parameters\n",
    "dataset_name = params[\"dataset_name\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Input files\n",
    "base_dir = os.path.abspath(os.path.join(os.getcwd(),\"../..\")) \n",
    "\n",
    "mapping_file = os.path.join(\n",
    "    base_dir,\n",
    "    dataset_name,\n",
    "    \"data\",\n",
    "    \"metadata\",\n",
    "    \"sample_annotations.tsv\")\n",
    "\n",
    "normalized_data_file = os.path.join(\n",
    "    base_dir,\n",
    "    dataset_name,\n",
    "    \"data\",\n",
    "    \"input\",\n",
    "    \"train_set_normalized.pcl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output file\n",
    "experiment_id_file = os.path.join(\n",
    "    base_dir, \n",
    "    dataset_name,\n",
    "    \"data\",\n",
    "    \"metadata\", \n",
    "    \"experiment_ids.txt\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get experiment ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>ml_data_source</th>\n",
       "      <th>description</th>\n",
       "      <th>nucleic_acid</th>\n",
       "      <th>medium</th>\n",
       "      <th>genotype</th>\n",
       "      <th>od</th>\n",
       "      <th>growth_setting_1</th>\n",
       "      <th>growth_setting_2</th>\n",
       "      <th>strain</th>\n",
       "      <th>temperature</th>\n",
       "      <th>treatment</th>\n",
       "      <th>additional_notes</th>\n",
       "      <th>variant_phenotype</th>\n",
       "      <th>abx_marker</th>\n",
       "      <th>biotic_int_lv_2</th>\n",
       "      <th>biotic_int_lv_1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>experiment</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>E-GEOD-46947</th>\n",
       "      <td>GSM1141730 1</td>\n",
       "      <td>GSM1141730_PA01_ZnO_PZO_.CEL</td>\n",
       "      <td>Pseudomonas aeruginosa PAO1 LB aerated  5 h wi...</td>\n",
       "      <td>RNA</td>\n",
       "      <td>LB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>planktonic</td>\n",
       "      <td>aerated</td>\n",
       "      <td>PAO1</td>\n",
       "      <td>37.0</td>\n",
       "      <td>1 mM ZnO nanoparticles</td>\n",
       "      <td>Grown for 5h</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E-GEOD-46947</th>\n",
       "      <td>GSM1141729 1</td>\n",
       "      <td>GSM1141729_PA01_none_PC_.CEL</td>\n",
       "      <td>Pseudomonas aeruginosa PAO1 LB aerated  5 h</td>\n",
       "      <td>RNA</td>\n",
       "      <td>LB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>planktonic</td>\n",
       "      <td>aerated</td>\n",
       "      <td>PAO1</td>\n",
       "      <td>37.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Grown for 5h</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E-GEOD-65882</th>\n",
       "      <td>GSM1608059 1</td>\n",
       "      <td>GSM1608059_Planktonic_1.CEL</td>\n",
       "      <td>PAO1 WT. Planktonic. Rep1</td>\n",
       "      <td>RNA</td>\n",
       "      <td>PBM plus 1 g / L glucose.</td>\n",
       "      <td>WT</td>\n",
       "      <td>0.26</td>\n",
       "      <td>Planktonic</td>\n",
       "      <td>Aerated</td>\n",
       "      <td>PAO1</td>\n",
       "      <td>37.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Grown shaking at 200rpm</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E-GEOD-65882</th>\n",
       "      <td>GSM1608060 1</td>\n",
       "      <td>GSM1608060_Planktonic_2.CEL</td>\n",
       "      <td>PAO1 WT. Planktonic. Rep2</td>\n",
       "      <td>RNA</td>\n",
       "      <td>PBM plus 1 g / L glucose.</td>\n",
       "      <td>WT</td>\n",
       "      <td>0.26</td>\n",
       "      <td>Planktonic</td>\n",
       "      <td>Aerated</td>\n",
       "      <td>PAO1</td>\n",
       "      <td>37.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Grown shaking at 200rpm</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E-GEOD-65882</th>\n",
       "      <td>GSM1608061 1</td>\n",
       "      <td>GSM1608061_Planktonic_3.CEL</td>\n",
       "      <td>PAO1 WT. Planktonic. Rep3</td>\n",
       "      <td>RNA</td>\n",
       "      <td>PBM plus 1 g / L glucose.</td>\n",
       "      <td>WT</td>\n",
       "      <td>0.26</td>\n",
       "      <td>Planktonic</td>\n",
       "      <td>Aerated</td>\n",
       "      <td>PAO1</td>\n",
       "      <td>37.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Grown shaking at 200rpm</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               sample_name                ml_data_source  \\\n",
       "experiment                                                 \n",
       "E-GEOD-46947  GSM1141730 1  GSM1141730_PA01_ZnO_PZO_.CEL   \n",
       "E-GEOD-46947  GSM1141729 1  GSM1141729_PA01_none_PC_.CEL   \n",
       "E-GEOD-65882  GSM1608059 1   GSM1608059_Planktonic_1.CEL   \n",
       "E-GEOD-65882  GSM1608060 1   GSM1608060_Planktonic_2.CEL   \n",
       "E-GEOD-65882  GSM1608061 1   GSM1608061_Planktonic_3.CEL   \n",
       "\n",
       "                                                    description nucleic_acid  \\\n",
       "experiment                                                                     \n",
       "E-GEOD-46947  Pseudomonas aeruginosa PAO1 LB aerated  5 h wi...          RNA   \n",
       "E-GEOD-46947        Pseudomonas aeruginosa PAO1 LB aerated  5 h          RNA   \n",
       "E-GEOD-65882                          PAO1 WT. Planktonic. Rep1          RNA   \n",
       "E-GEOD-65882                          PAO1 WT. Planktonic. Rep2          RNA   \n",
       "E-GEOD-65882                          PAO1 WT. Planktonic. Rep3          RNA   \n",
       "\n",
       "                                  medium genotype    od growth_setting_1  \\\n",
       "experiment                                                                 \n",
       "E-GEOD-46947                          LB      NaN   NaN       planktonic   \n",
       "E-GEOD-46947                          LB      NaN   NaN       planktonic   \n",
       "E-GEOD-65882  PBM plus 1 g / L glucose.        WT  0.26      Planktonic    \n",
       "E-GEOD-65882  PBM plus 1 g / L glucose.        WT  0.26      Planktonic    \n",
       "E-GEOD-65882  PBM plus 1 g / L glucose.        WT  0.26      Planktonic    \n",
       "\n",
       "             growth_setting_2 strain  temperature               treatment  \\\n",
       "experiment                                                                  \n",
       "E-GEOD-46947          aerated   PAO1         37.0  1 mM ZnO nanoparticles   \n",
       "E-GEOD-46947          aerated   PAO1         37.0                     NaN   \n",
       "E-GEOD-65882          Aerated   PAO1         37.0                     NaN   \n",
       "E-GEOD-65882          Aerated   PAO1         37.0                     NaN   \n",
       "E-GEOD-65882          Aerated   PAO1         37.0                     NaN   \n",
       "\n",
       "                     additional_notes variant_phenotype abx_marker  \\\n",
       "experiment                                                           \n",
       "E-GEOD-46947             Grown for 5h               NaN        NaN   \n",
       "E-GEOD-46947             Grown for 5h               NaN        NaN   \n",
       "E-GEOD-65882  Grown shaking at 200rpm               NaN        NaN   \n",
       "E-GEOD-65882  Grown shaking at 200rpm               NaN        NaN   \n",
       "E-GEOD-65882  Grown shaking at 200rpm               NaN        NaN   \n",
       "\n",
       "             biotic_int_lv_2 biotic_int_lv_1  \n",
       "experiment                                    \n",
       "E-GEOD-46947             NaN             NaN  \n",
       "E-GEOD-46947             NaN             NaN  \n",
       "E-GEOD-65882             NaN             NaN  \n",
       "E-GEOD-65882             NaN             NaN  \n",
       "E-GEOD-65882             NaN             NaN  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read in metadata\n",
    "metadata = pd.read_table(\n",
    "    mapping_file, \n",
    "    header=0, \n",
    "    sep='\\t', \n",
    "    index_col=0)\n",
    "\n",
    "metadata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>ml_data_source</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>experiment</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>E-GEOD-46947</th>\n",
       "      <td>GSM1141730 1</td>\n",
       "      <td>GSM1141730_PA01_ZnO_PZO_.CEL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E-GEOD-46947</th>\n",
       "      <td>GSM1141729 1</td>\n",
       "      <td>GSM1141729_PA01_none_PC_.CEL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E-GEOD-65882</th>\n",
       "      <td>GSM1608059 1</td>\n",
       "      <td>GSM1608059_Planktonic_1.CEL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E-GEOD-65882</th>\n",
       "      <td>GSM1608060 1</td>\n",
       "      <td>GSM1608060_Planktonic_2.CEL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E-GEOD-65882</th>\n",
       "      <td>GSM1608061 1</td>\n",
       "      <td>GSM1608061_Planktonic_3.CEL</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               sample_name                ml_data_source\n",
       "experiment                                              \n",
       "E-GEOD-46947  GSM1141730 1  GSM1141730_PA01_ZnO_PZO_.CEL\n",
       "E-GEOD-46947  GSM1141729 1  GSM1141729_PA01_none_PC_.CEL\n",
       "E-GEOD-65882  GSM1608059 1   GSM1608059_Planktonic_1.CEL\n",
       "E-GEOD-65882  GSM1608060 1   GSM1608060_Planktonic_2.CEL\n",
       "E-GEOD-65882  GSM1608061 1   GSM1608061_Planktonic_3.CEL"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "map_experiment_sample = metadata[['sample_name', 'ml_data_source']]\n",
    "map_experiment_sample.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 135 experiments in the compendium\n"
     ]
    }
   ],
   "source": [
    "experiment_ids = np.unique(np.array(map_experiment_sample.index))\n",
    "print(\"There are {} experiments in the compendium\".format(len(experiment_ids)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get sample ids from gene expression data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Gene_symbol</th>\n",
       "      <th>PA0001</th>\n",
       "      <th>PA0002</th>\n",
       "      <th>PA0003</th>\n",
       "      <th>PA0004</th>\n",
       "      <th>PA0005</th>\n",
       "      <th>PA0006</th>\n",
       "      <th>PA0007</th>\n",
       "      <th>PA0008</th>\n",
       "      <th>PA0009</th>\n",
       "      <th>PA0010</th>\n",
       "      <th>...</th>\n",
       "      <th>PA5561</th>\n",
       "      <th>PA5562</th>\n",
       "      <th>PA5563</th>\n",
       "      <th>PA5564</th>\n",
       "      <th>PA5565</th>\n",
       "      <th>PA5566</th>\n",
       "      <th>PA5567</th>\n",
       "      <th>PA5568</th>\n",
       "      <th>PA5569</th>\n",
       "      <th>PA5570</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>05_PA14000-4-2_5-10-07_S2.CEL</th>\n",
       "      <td>0.853357</td>\n",
       "      <td>0.725280</td>\n",
       "      <td>0.640617</td>\n",
       "      <td>0.811465</td>\n",
       "      <td>0.694460</td>\n",
       "      <td>0.533958</td>\n",
       "      <td>0.158865</td>\n",
       "      <td>0.889579</td>\n",
       "      <td>0.884945</td>\n",
       "      <td>0.176558</td>\n",
       "      <td>...</td>\n",
       "      <td>0.466871</td>\n",
       "      <td>0.702785</td>\n",
       "      <td>0.790965</td>\n",
       "      <td>0.893249</td>\n",
       "      <td>0.789939</td>\n",
       "      <td>0.164157</td>\n",
       "      <td>0.970470</td>\n",
       "      <td>0.887472</td>\n",
       "      <td>0.900484</td>\n",
       "      <td>0.880012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54375-4-05.CEL</th>\n",
       "      <td>0.778790</td>\n",
       "      <td>0.767873</td>\n",
       "      <td>0.614859</td>\n",
       "      <td>0.907865</td>\n",
       "      <td>0.398800</td>\n",
       "      <td>0.460849</td>\n",
       "      <td>0.113876</td>\n",
       "      <td>0.761351</td>\n",
       "      <td>0.801740</td>\n",
       "      <td>0.222709</td>\n",
       "      <td>...</td>\n",
       "      <td>0.352020</td>\n",
       "      <td>0.694387</td>\n",
       "      <td>0.733186</td>\n",
       "      <td>0.639074</td>\n",
       "      <td>0.681204</td>\n",
       "      <td>0.110301</td>\n",
       "      <td>0.619554</td>\n",
       "      <td>0.747656</td>\n",
       "      <td>0.749893</td>\n",
       "      <td>0.805374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AKGlu_plus_nt_7-8-09_s1.CEL</th>\n",
       "      <td>0.789155</td>\n",
       "      <td>0.729508</td>\n",
       "      <td>0.725913</td>\n",
       "      <td>0.718989</td>\n",
       "      <td>0.530160</td>\n",
       "      <td>0.466327</td>\n",
       "      <td>0.079507</td>\n",
       "      <td>0.731643</td>\n",
       "      <td>0.827707</td>\n",
       "      <td>0.241847</td>\n",
       "      <td>...</td>\n",
       "      <td>0.392405</td>\n",
       "      <td>0.700352</td>\n",
       "      <td>0.773422</td>\n",
       "      <td>0.791118</td>\n",
       "      <td>0.931585</td>\n",
       "      <td>0.172570</td>\n",
       "      <td>0.797148</td>\n",
       "      <td>0.753785</td>\n",
       "      <td>0.856253</td>\n",
       "      <td>0.811099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>anaerobic_NO3_1.CEL</th>\n",
       "      <td>0.716320</td>\n",
       "      <td>0.585079</td>\n",
       "      <td>0.390211</td>\n",
       "      <td>0.193248</td>\n",
       "      <td>0.279456</td>\n",
       "      <td>0.301781</td>\n",
       "      <td>0.513547</td>\n",
       "      <td>0.342051</td>\n",
       "      <td>0.415668</td>\n",
       "      <td>0.125914</td>\n",
       "      <td>...</td>\n",
       "      <td>0.398308</td>\n",
       "      <td>0.419574</td>\n",
       "      <td>0.593955</td>\n",
       "      <td>0.527203</td>\n",
       "      <td>0.706524</td>\n",
       "      <td>0.205510</td>\n",
       "      <td>0.504767</td>\n",
       "      <td>0.105662</td>\n",
       "      <td>0.363409</td>\n",
       "      <td>0.544780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>anaerobic_NO3_2.CEL</th>\n",
       "      <td>0.658015</td>\n",
       "      <td>0.592172</td>\n",
       "      <td>0.410331</td>\n",
       "      <td>0.245504</td>\n",
       "      <td>0.312028</td>\n",
       "      <td>0.305852</td>\n",
       "      <td>0.513499</td>\n",
       "      <td>0.336723</td>\n",
       "      <td>0.334226</td>\n",
       "      <td>0.162965</td>\n",
       "      <td>...</td>\n",
       "      <td>0.407801</td>\n",
       "      <td>0.478697</td>\n",
       "      <td>0.571460</td>\n",
       "      <td>0.473054</td>\n",
       "      <td>0.669643</td>\n",
       "      <td>0.155548</td>\n",
       "      <td>0.562927</td>\n",
       "      <td>0.049738</td>\n",
       "      <td>0.388931</td>\n",
       "      <td>0.548814</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 5549 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene_symbol                      PA0001    PA0002    PA0003    PA0004  \\\n",
       "05_PA14000-4-2_5-10-07_S2.CEL  0.853357  0.725280  0.640617  0.811465   \n",
       "54375-4-05.CEL                 0.778790  0.767873  0.614859  0.907865   \n",
       "AKGlu_plus_nt_7-8-09_s1.CEL    0.789155  0.729508  0.725913  0.718989   \n",
       "anaerobic_NO3_1.CEL            0.716320  0.585079  0.390211  0.193248   \n",
       "anaerobic_NO3_2.CEL            0.658015  0.592172  0.410331  0.245504   \n",
       "\n",
       "Gene_symbol                      PA0005    PA0006    PA0007    PA0008  \\\n",
       "05_PA14000-4-2_5-10-07_S2.CEL  0.694460  0.533958  0.158865  0.889579   \n",
       "54375-4-05.CEL                 0.398800  0.460849  0.113876  0.761351   \n",
       "AKGlu_plus_nt_7-8-09_s1.CEL    0.530160  0.466327  0.079507  0.731643   \n",
       "anaerobic_NO3_1.CEL            0.279456  0.301781  0.513547  0.342051   \n",
       "anaerobic_NO3_2.CEL            0.312028  0.305852  0.513499  0.336723   \n",
       "\n",
       "Gene_symbol                      PA0009    PA0010    ...       PA5561  \\\n",
       "05_PA14000-4-2_5-10-07_S2.CEL  0.884945  0.176558    ...     0.466871   \n",
       "54375-4-05.CEL                 0.801740  0.222709    ...     0.352020   \n",
       "AKGlu_plus_nt_7-8-09_s1.CEL    0.827707  0.241847    ...     0.392405   \n",
       "anaerobic_NO3_1.CEL            0.415668  0.125914    ...     0.398308   \n",
       "anaerobic_NO3_2.CEL            0.334226  0.162965    ...     0.407801   \n",
       "\n",
       "Gene_symbol                      PA5562    PA5563    PA5564    PA5565  \\\n",
       "05_PA14000-4-2_5-10-07_S2.CEL  0.702785  0.790965  0.893249  0.789939   \n",
       "54375-4-05.CEL                 0.694387  0.733186  0.639074  0.681204   \n",
       "AKGlu_plus_nt_7-8-09_s1.CEL    0.700352  0.773422  0.791118  0.931585   \n",
       "anaerobic_NO3_1.CEL            0.419574  0.593955  0.527203  0.706524   \n",
       "anaerobic_NO3_2.CEL            0.478697  0.571460  0.473054  0.669643   \n",
       "\n",
       "Gene_symbol                      PA5566    PA5567    PA5568    PA5569  \\\n",
       "05_PA14000-4-2_5-10-07_S2.CEL  0.164157  0.970470  0.887472  0.900484   \n",
       "54375-4-05.CEL                 0.110301  0.619554  0.747656  0.749893   \n",
       "AKGlu_plus_nt_7-8-09_s1.CEL    0.172570  0.797148  0.753785  0.856253   \n",
       "anaerobic_NO3_1.CEL            0.205510  0.504767  0.105662  0.363409   \n",
       "anaerobic_NO3_2.CEL            0.155548  0.562927  0.049738  0.388931   \n",
       "\n",
       "Gene_symbol                      PA5570  \n",
       "05_PA14000-4-2_5-10-07_S2.CEL  0.880012  \n",
       "54375-4-05.CEL                 0.805374  \n",
       "AKGlu_plus_nt_7-8-09_s1.CEL    0.811099  \n",
       "anaerobic_NO3_1.CEL            0.544780  \n",
       "anaerobic_NO3_2.CEL            0.548814  \n",
       "\n",
       "[5 rows x 5549 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "normalized_data = pd.read_table(\n",
    "    normalized_data_file,\n",
    "    header=0,\n",
    "    sep='\\t',\n",
    "    index_col=0).T\n",
    "\n",
    "normalized_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_ids_with_gene_expression = list(normalized_data.index)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get samples belonging to selected experiment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 107 experiments with gene expression data\n"
     ]
    }
   ],
   "source": [
    "experiment_ids_with_gene_expression = []\n",
    "\n",
    "for experiment_id in experiment_ids:\n",
    "    selected_metadata = metadata.loc[experiment_id]\n",
    "    \n",
    "    #print(\"There are {} samples in experiment {}\".format(selected_metadata.shape[0], experiment_id))\n",
    "    \n",
    "    sample_ids = list(selected_metadata['ml_data_source'])\n",
    "    \n",
    "    if any(x in sample_ids_with_gene_expression for x in sample_ids):\n",
    "        experiment_ids_with_gene_expression.append(experiment_id)\n",
    "        \n",
    "print('There are {} experiments with gene expression data'.format(len(experiment_ids_with_gene_expression)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>experiment_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>E-GEOD-10030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>E-GEOD-10065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>E-GEOD-10362</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>E-GEOD-10604</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>E-GEOD-11544</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  experiment_id\n",
       "0  E-GEOD-10030\n",
       "1  E-GEOD-10065\n",
       "2  E-GEOD-10362\n",
       "3  E-GEOD-10604\n",
       "4  E-GEOD-11544"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "experiment_ids_with_gene_expression_df = pd.DataFrame(experiment_ids_with_gene_expression, columns=['experiment_id'])\n",
    "experiment_ids_with_gene_expression_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save simulated data\n",
    "experiment_ids_with_gene_expression_df.to_csv(experiment_id_file, sep='\\t')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:batch_effects]",
   "language": "python",
   "name": "conda-env-batch_effects-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
